tensorflow/core/framework/resource_mgr.cc

```diff
@@ -107,14 +107,30 @@ ResourceMgr::ResourceMgr(const string& default_container)
 ResourceMgr::~ResourceMgr() { Clear(); }
 
 void ResourceMgr::Clear() {
-  mutex_lock l(mu_);
-  for (const auto& p : containers_) {
-    for (const auto& q : *p.second) {
+  // In some cases (e.g. DatasetConstructorTest.testSharedName),
+  // `ResourceMgr::Clear` will trigger the call of `ResourceMgr::Cleanup`. It
+  // will cause the deadlock if `ResourceMgr::Clear` still holds the lock. To
+  // avoid this issue, we copy the content of `containers_` to a temporary local
+  // variable under the mutex and then clear the resources in the containers
+  // without holding the lock.
+  std::vector<Container*> copied_containers;
```

Comment:
Why do you create a copy here? When I suggested you move it, I meant the following:

```
void ResourceMgr::Clear() {
  std::unordered_map<string, Container*> tmp_containers;
  {
    mutex_lock l(mu_);
    tmp_containers = std::move(containers_);
  }
  for (const auto& p : tmp_containers) {
    for (const auto& q : *p.second) {
      q.second->Unref();
    }
    delete p.second;
  }
  tmp_containers.clear();
}
```

Reply:
Yes, this is better. Will change it. Thanks for the detailed codes!